### PR TITLE
Fix for logic error in TestObjectSizeResponses

### PR DESCRIPTION
**Pull request by mkenigs**
**What this PR does / why we need it**:
TestObjectSizeResponses should be testing when the error "etcdserver: request is too large" is thrown, but DeploymentMegabyteSize isn't large enough to throw the error. This isn't caught because TestObjectSizeResponses assumes that an error will be thrown.

**Which issue(s) this PR fixes**
Helps #47668

**Special notes for your reviewer**:
I changed DeploymentMegabyteSize to 130000 because it's large enough to throw the etcd error but not so large that it throws the gRPC error.
@sttts @anubhakushwaha

### CODE REVIEW

File: test/integration/master/synthetic_master_test.go [hunk 1]
```diff
@@ -300,7 +300,7 @@ func TestObjectSizeResponses(t *testing.T) {
 
 	client := clientsetv1.NewForConfigOrDie(&restclient.Config{Host: s.URL, ContentConfig: restclient.ContentConfig{GroupVersion: testapi.Groups[api.GroupName].GroupVersion()}})
 
-	const DeploymentMegabyteSize = 100000
+	const DeploymentMegabyteSize = 130000
```
Comment (dixudx) [inline test/integration/master/synthetic_master_test.go]: Why changing this value?

Comment (sttts) [reply test/integration/master/synthetic_master_test.go]: Maybe add another test (or check in the actual test) which verifies this number, i.e. that the body has the right size?

Comment (roycaihw) [reply test/integration/master/synthetic_master_test.go]: For `constructBody`, I think you could pad zero to the number when printing to make it easier to calculate the size of created object, like:

```
 for i := 0; i < size; i++ { 
 	key := val + fmt.Sprintf("%010d", i) // len(key) = len(val) + 10
 	labelsMap[key] = val 
 } 
```

**Comment by mkenigs** [comment]
Most recent commit pads zero to the number and makes a few other changes to make it easier to calculate the size of the body. The tests are changed to 1.5 MB and 2 MB because those are the sizes that cause the etcd and gRPC errors, respectively. The previous version labeled them as 1 MB and 2 MB, but that wasn't accurate due to the way constructBody was written.

File: test/integration/master/synthetic_master_test.go [hunk 2]
```diff
@@ -300,30 +300,34 @@ func TestObjectSizeResponses(t *testing.T) {
 
 	client := clientsetv1.NewForConfigOrDie(&restclient.Config{Host: s.URL, ContentConfig: restclient.ContentConfig{GroupVersion: testapi.Groups[api.GroupName].GroupVersion()}})
 
-	const DeploymentMegabyteSize = 100000
-	const DeploymentTwoMegabyteSize = 1000000
+	const labelString = ""
+	const finalizerString = "sample/sample"
+	// for given labelString and finalizerString, size of the object returned by constructBody is increased by factor*(size passed)
+	const factor = 16
```
Comment (roycaihw) [inline test/integration/master/synthetic_master_test.go]: Why is factor 16? Could you add documentation on that?

**Comment by roycaihw** [comment]
> Still working on another test to check if the size of the object created by constructBody is correct - would that be helpful?

Yes please. Either another test or verify in the actual test will be helpful.

File: test/integration/master/synthetic_master_test.go (hunk 2 above)
Comment (mkenigs) [reply test/integration/master/synthetic_master_test.go]: For a label or annotation, each iteration adds a key of length 10 and a value of length length(labelString). I'm not sure why, but that has a total size of 16.
For a finalizer, each iteration appends finalizerString to a []string{}. Again not sure why, but for length(finalizerString)=13, that has a total size of 16.

**Comment by mkenigs** [comment]
Anything else I need to do on this? @roycaihw

**Review by roycaihw — COMMENTED** [review]
Please see comments below about the constants. We could document the tests to make it clear where these constants (`deploymentObjectSize`, `baseLabelSize`, `factor`) come from, but I'm feeling these numbers are subject to API change of `appsv1.Deployment` implementation (not sure how likely that's going to happen), which might make the test hard to maintain.

Maybe just verifying the body size in `TestObjectSizeResponses` is greater than 1.5/2 MB would be sufficient. But I'm also okay to LGTM the current separate test if we document the constants well.

@sttts @dixudx Any suggestions?

**Comment by mkenigs** [comment]
Per https://github.com/kubernetes/kubernetes/issues/47668#issuecomment-397029553, ultimately it sounds like there should just be one error for an object larger than 1.5 MB, which would simplify this test and all the constants. It would be easier to guarantee an object >1.5MB rather than making sure it's between 1.5 and 2.

File: test/integration/master/synthetic_master_test.go [hunk 3]
```diff
@@ -334,6 +340,70 @@ func TestObjectSizeResponses(t *testing.T) {
...
+				t.Errorf("could not encode deployment object: %s", err.Error())
+			} else {
+				var expectedSize int
+				switch r.field {
+				case "labels":
+					expectedSize = baseLabelSize + factor*r.size
+				case "annotations":
+					expectedSize = baseAnnotationsSize + factor*r.size
+				case "finalizers":
+					expectedSize = baseFinalizersSize + factor*r.size
```
[the remark this answers is not in the document]
Comment (mkenigs) [reply test/integration/master/synthetic_master_test.go]: Agreed, changed that

**Comment by dixudx** [comment]
> We could document the tests to make it clear where these constants (deploymentObjectSize, baseLabelSize, factor) come from,

@roycaihw Agree, that will be pretty good. 

> Maybe just verifying the body size in TestObjectSizeResponses is greater than 1.5/2 MB would be sufficient.

@roycaihw @mkenigs I think this is the right direction. We should make it clear what this test is going to do instead of many unclear numbers and formulas.

@mkenigs Please rebase the commits for better review. Thanks for your contribution.

**Comment by mkenigs** [comment]
@roycaihw was trying to fix #47668 before changing these tests, but not quite sure where to make the fix. Looks like apiserver could be changed to properly handle the etcd and gRPC errors, or maybe validation could be done on all objects to make sure they don't exceed a max size, but I'm not sure what would be best or where those changes would need to be made.

I could also just simplify the test for the current behavior - maybe ignore the size of the object except for the added label/finalizer/annotation to eliminate some of the constants.

**Comment by roycaihw** [comment]
@mkenigs fix the test can be a good starting point. Ideally we want a unified error message instead of `apiserver received an error that is not an metav1.Status` https://github.com/kubernetes/kubernetes/issues/47668#issuecomment-397029553

> apiserver could be changed to properly handle the etcd and gRPC errors, or maybe validation could be done on all objects to make sure they don't exceed a max size

I'm in favor of validation to save some bandwidth. but note that the limit is on the wired form (protobuf by default), so the validation have to happen pretty late in our stack to know the size of the request and maybe we are not saving much bandwidth

we can have the error handling regardless of having validation or not. @sttts @dixudx Any suggestions?

**Comment by mkenigs** [comment]
> Maybe just verifying the body size in `TestObjectSizeResponses` is greater than 1.5/2 MB would be sufficient.
@roycaihw changed the test to check that the object is within .5 MB of what is expected, instead of calculating exactly how big it should be

**Comment by mkenigs** [comment]
/remove-area sig/cloud-provider sig/cluster-lifecycle sig/contributor-experience sig/gcp sig/instrumentation sig/release sig/scheduling

**Pull request closed without merging**